File: Napakalaki-Game(PDOO)/NapakalakiJava/src/napakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author joaquin
 */
public class Napakalaki {
    
    private static final Napakalaki instance = new Napakalaki();
    
    private ArrayList<Player> players = new ArrayList();
    private Player currentPlayer;
    private CardDealer dealer = CardDealer.getInstance();
    private Monster currentMonster;
    
    
    private Napakalaki(){
    
    }
    
    private void initPlayers(ArrayList<String>names){
      for (String name:names){
        players.add(new Player(name));
      }
    }
    
    private Player nextPlayer(){
        Random generator=new Random();
        Player jugador;
        int pos;
        if(currentPlayer==null){
           int numero= (generator.nextInt(players.size()));
           jugador=players.get(numero);
        }else{
           pos=players.indexOf(currentPlayer);
           if(pos==players.size()-1){
               pos=0;
           }else{
               pos++;
           }
           jugador=players.get(pos);
        }
        currentPlayer=jugador;
    return jugador;
    }
    
    private boolean nextTurnAllowed(){
        boolean puede;
        if(currentPlayer==null)
            puede=true;
        else
           puede=currentPlayer.validState();
    return puede;
    }
    
    private void setEnemies(){
       Player jugador;
       int numero;
       for (Player player:players){
             do{ 
                 Random generator=new Random();
                 numero= generator.nextInt(players.size());
                 jugador=players.get(numero);
             }while(jugador==player);
             player.setEnemy(player);
      }
    }
    
    public static Napakalaki getInstance(){
    return Napakalaki.instance;
    }
    
    public CombatResult developCombat(){
        CombatResult resultado;
        Cultist carta;
        CultistPlayer nuevo;
        resultado= currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        if(resultado==CombatResult.LOSEANDCONVERT){
          carta=dealer.nextCultist();
          nuevo =new CultistPlayer(currentPlayer,carta);
          for (Player player:players){
             if(player.getEnemy()==currentPlayer)
             player.setEnemy(nuevo);
          }
          players.remove(currentPlayer);
          currentPlayer=nuevo;
          players.add(currentPlayer);
          
        }
    return resultado;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure>treasures){
          for(Treasure treasure:treasures){
              currentPlayer.discardVisibleTreasure(treasure);
              dealer.giveTreasureBack(treasure);
          }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure>treasures){
           for(Treasure treasure:treasures){
              currentPlayer.discardHiddenTreasure(treasure);
              dealer.giveTreasureBack(treasure);
          }       
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure>treasures){
        for(Treasure t:treasures){
           currentPlayer.makeTreasureVisible(t);
           
        }
            
    }
    
    public void initGame(ArrayList<String> names){
        initPlayers(names);
        setEnemies();
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
    return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
    return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean dead;
        boolean stateOK = nextTurnAllowed();
        if(stateOK){
           currentMonster=dealer.nextMonster();
           currentPlayer = nextPlayer();
           dead= currentPlayer.isDead();
           if(dead){
              currentPlayer.initTreasures();
             }
        
        }
    return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        boolean salida= false;
        if(result==CombatResult.WINGAME)
            salida=true;
    return salida;
    }
}

File: Napakalaki-Game(PDOO)/NapakalakiRuby/lib/napakalaki.rb
# -*- coding: utf-8 -*-
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require 'singleton'
require_relative 'player.rb'
require_relative 'card_dealer.rb'
require_relative 'combat_result.rb'
require_relative 'cultist_player.rb'
class Napakalaki
  include Singleton
  
private
  
  def initialize()
    @currentPlayer=nil
    @players = Array.new
    @dealer = CardDealer.instance
    @currentMonster=nil
    
  end
  
  def initPlayers(names)
      names.each do |name|
      jugador = Player.new(name)
      
      @players.push(jugador) 
    end

  end
  
  def nextPlayer()
        if(@currentPlayer==nil)
           r = Random.new
           numero = r.rand(0...@players.size())
           jugador= @players.at(numero)
        else
           pos=@players.index(@currentPlayer)
           if(pos==@players.size()-1)
               pos=0
           else
               pos=pos+1
           end
           jugador=@players.at(pos)
        end
        @currentPlayer=jugador
    return jugador
  end
  
  def nextTurnAllowed()
        if(@currentPlayer==nil)
            puede=true
        else
           puede=@currentPlayer.validState()
        end
    return puede
  end
  def setEnemies()
           @players.each do |player|
             loop do
                 r = Random.new
                 numero = r.rand(0..@players.size()) 
                 jugador=@players.at(numero)
             break if jugador==player
             end
             player.setEnemy(player)
             
           end
  end
  
  public
  
  def developCombat()
        resultado= @currentPlayer.combat(@currentMonster)
        @dealer.giveMonsterBack(@currentMonster)
        if(resultado==NapakalakiGame::CombatResult::LOSEANDCONVERT)
          carta=@dealer.nextCultist()
          nuevo =CultistPlayer.new(@currentPlayer,carta)
          @players.each do|player|
             if(player.enemy==@currentPlayer)
             player.setEnemy(nuevo)
             end
          end
          @players.delete(@currentPlayer)
          @currentPlayer=nuevo
          @players.push(@currentPlayer)
        end
    return resultado
  end
  
  def discardVisibleTreasures(treasures)
      treasures.each do |treasure|
              @currentPlayer.discardVisibleTreasure(treasure)
              @dealer.giveTreasureBack(treasure)
      end
  end
  
  def discardHiddenTreasures(treasures)
    treasures.each do |treasure|
              @currentPlayer.discardHiddenTreasure(treasure)
              @dealer.giveTreasureBack(treasure)
      end
  end
  
  def makeTreasuresVisible(treasures)
    treasures.each do |t|
           @currentPlayer.makeTreasureVisible(t)
           
    end
  end
  
  def initGame(names)
       initPlayers(names)
       setEnemies()
       @dealer.initCards()
       nextTurn()
  end
  
  def getCurrentPlayer()
    return @currentPlayer
  end
  
  def getCurrentMonster()
    return @currentMonster
  end
  
  def nextTurn()
     
        stateOK = nextTurnAllowed()
        if(stateOK)
           @currentMonster=@dealer.nextMonster()
           @currentPlayer = nextPlayer()
           dead= @currentPlayer.isDead()
           if(dead)
              @currentPlayer.initTreasures()
           end
        
        end
    return stateOK
  end
  
  def endOfGame(result)
    salida= false
        if(result==CombatResult::WINGAME)
            salida=true
        end
    return salida
  end
end
